src/CBLURLEndpointListener.cc

```diff
@@ -0,0 +1,216 @@
+            tls.tlsCallbackContext = this;
+            tls.certAuthCallback = [](C4Listener *listener, C4Slice clientCertData, void *context) -> bool {
+                auto me = reinterpret_cast<CBLURLEndpointListener*>(context);
+                return me->_conf.authenticator->certCallback(me->_conf.context, clientCertData);
+            };
+        } else {
+            // user/password
+            c4config.httpAuthCallback = [](C4Listener* listener, C4Slice authHeader, void* context) -> bool {
+                slice header{authHeader};
+                auto space = header.findByte(' ');
+                if (!space) return false;
+
+                else header = header.from(space + 1);
+                while ( !header.empty() && header[0] == ' ' )
+                    header = header.from(1);
+                alloc_slice decoded = fleece::base64::decode(header);
+                auto colon = decoded.findByte(':');
+                if (!colon) return false;
+
+                slice usr = decoded.upTo(colon);
+                slice pwd = decoded.from(colon + 1);
+
+                auto me = reinterpret_cast<CBLURLEndpointListener*>(context);
+                Assert(me->_c4listener == listener);
+                return me->_conf.authenticator->pswCallback(me->_conf.context, usr, pwd);
+            };
+            c4config.callbackContext = this;
+        }
+    }
+
+    std::unique_ptr<C4Listener> c4listener{new C4Listener(c4config)};
+    auto ret = [&](bool succ) -> bool {
+        if (succ) {
+            _c4listener = c4listener.release();
+        }
+        return succ;
+    };
+
+    CBLDatabase* cblDb = _conf.collections[0]->database();
+    bool succ = true;
+    cblDb->c4db()->useLocked([&](C4Database* db) {
+        slice dbname = db->getName();
+        if ( (succ = c4listener->shareDB(dbname, db)) ) {
+            for (unsigned i = 0; i < _conf.collectionCount; ++i) {
+                _conf.collections[i]->useLocked([&](C4Collection* coll) {
+                    succ = c4listener->shareCollection(dbname, coll);
+                });
+                if (!succ) break;
+            }
+        }
+    });
+
+    return ret(succ);
+}
+
+void CBLURLEndpointListener::stop() {
+    std::scoped_lock lock(_mutex);
+    if (_c4listener) {
+        delete _c4listener;
+        _c4listener = nullptr;
+    }
+}
+
+std::string CBLURLEndpointListener::dumpConfig() {
+    return {};
+}
+
+CBLTLSIdentity* CBLURLEndpointListener::effectiveTLSIdentity(bool persistent) {
+    if (_conf.disableTLS)
+        return nullptr;
+
+    if (!_effectiveTLSIdentity)
+        _effectiveTLSIdentity = _conf.tlsIdentity ? _conf.tlsIdentity : anonymousTLSIdentity(persistent).get();
+
+    return _effectiveTLSIdentity;
+}
+
+Retained<CBLTLSIdentity> CBLURLEndpointListener::anonymousTLSIdentity(bool persistent) {
+    Retained<CBLTLSIdentity> identity;
+    alloc_slice label;
+    std::unique_ptr<CBLKeyPair, void(*)(CBLKeyPair*)> keypair{
```

Comment:
Can you explain why not using Retained<CBLKeyPair>?

Reply:
This is because Retained will add a ref count. For pointers returned from C API, they already have ref count 1. It's to be released by XYZ_Release.

Reply:
The public C functions shouldn't be use directly in here. In general the structure of the code should be :

* Public C API uses CBL-C's C++ internal implementation
    * Catch exception from C++ and convert and set the error to out error.
* C++ internal implementation uses LiteCore C++ API.
* C++ internal implementation uses CBL-C's C++ internal implementation as needed.
    * C++ internal implementation throws exception when an error occurs.

Reply:
I changed to use c++ function which returns a bare pointer, I can retain it and no need to use unique_ptr